port/unix/omrsysinfo.c

```diff
@@ -5640,7 +5702,149 @@ readCgroupFile(struct OMRPortLibrary *portLibrary, int pid, BOOLEAN inContainer,
 		*cgroupEntryList = cgEntryList;
 		if (NULL != availableSubsystems) {
 			*availableSubsystems = available;
-			Trc_PRT_readCgroupFile_available_subsystems(available);
+			Trc_PRT_populateCgroupEntryList_available_subsystems(1, available);
+		}
+	}
+
+	return rc;
+}
+
+/**
+ * Reads /proc/<pid>/cgroup to get the name of the cgroup to which the given process
+ * belongs, then the $MOUNT_POINT/cgroupName/cgroup.controllers file to get the
+ * enabled subsystems.
+ *
+ * @param[in] portLibrary pointer to OMRPortLibrary
+ * @param[in] pid process id
+ * @param[out] cgroupEntryList pointer to OMRCgroupEntry *. On successful return, *cgroupEntry
+ * points to a circular linked list. Each element of the list is populated based on the contents
+ * of /proc/<pid>/cgroup file.
+ * @param[out] availableSubsystems on successful return, contains bitwise-OR of flags of type OMR_CGROUP_SUBSYSTEMS_*
+ * indicating the subsystems available for use
+ *
+ * returns 0 on success, negative code on error
+ */
+static int32_t
+populateCgroupEntryListV2(struct OMRPortLibrary *portLibrary, int pid, OMRCgroupEntry **cgroupEntryList, uint64_t *availableSubsystems)
+{
+	char cgroupFilePath[PATH_MAX];
+	char controllerFilePath[PATH_MAX];
+	char buffer[PATH_MAX];
+	uintptr_t requiredSize = 0;
+	FILE *cgroupFile = NULL;
+	FILE *controllerFile = NULL;
+	OMRCgroupEntry *cgEntryList = NULL;
+	uint64_t available = 0;
+	int32_t rc = 0;
+
+	Assert_PRT_true(NULL != cgroupEntryList);
+
+	requiredSize = portLibrary->str_printf(portLibrary, NULL, 0, "/proc/%d/cgroup", pid);
+	Assert_PRT_true(requiredSize <= PATH_MAX);
+	portLibrary->str_printf(portLibrary, cgroupFilePath, sizeof(cgroupFilePath), "/proc/%d/cgroup", pid);
```

Comment:
This could be improved (in a future pull request): there's no reason to call `str_printf()` twice if we are going to insist the result fit the the buffer we already have.
The assertion (that would go away) should use `sizeof(cgroupFilePath)` instead of `PATH_MAX`.